docs/design-documents/platform/pin-names/pin_names.md

```diff
@@ -0,0 +1,198 @@
+
+    // SPI signals are aliased as below  
+    
+    SPI_CS = D10,   // SPI CS  
+    SPI_MOSI = D11, // SPI MOSI  
+    SPI_MISO = D12, // SPI MISO  
+    SPI_SCK = D13,  // SPI SCK
+
+### Custom/Vendor Header Pins
+
+**Custom/Vendor Connector I2C Definition**
+
+All I2C definitions for a new connector should be defined as follows:
+
+    // ConnectorName is the name of the specific connector you are implementing signals for here  
+    // Px_xx relates to the processor pin connected to the custom/vendor connector pin  
+    
+    ConnectorName_I2C0_SCL  = Px_xx,  // Connector I2C0 SCL  
+    ConnectorName_I2C0_SDA  = Px_xx,  // Connector I2C0 SDA  
+    ConnectorName_I2C1_SCL  = Px_xx,  // Connector I2C1 SCL  
+    ConnectorName_I2C1_SDA  = Px_xx,  // Connector I2C1 SDA  
+    .  
+    .  
+    .  
+    ConnectorName_I2CN_SCL  = Px_xx,  // Connector I2CN SCL  
+    ConnectorName_I2CN_SDA  = Px_xx,  // Connector I2CN SDA  
+
+**Custom/Vendor Connector SPI Definition**
+
+All SPI definitions for a new connector should be defined as follows:
+
+    // ConnectorName is the name of the specific connector you are implementing signals for here  
+    // Px_xx relates to your processors pin connected to the custom/vendor connector pin
+    
+    ConnectorName_SPI0_CS   = Px_xx,  // Connector SPI0 CS  
+    ConnectorName_SPI0_MOSI = Px_xx,  // Connector SPI0 MOSI  
+    ConnectorName_SPI0_MISO = Px_xx,  // Connector SPI0 MISO  
+    ConnectorName_SPI0_SCK  = Px_xx,  // Connector SPI0 SCK  
+    
+    ConnectorName_SPI1_CS   = Px_xx,  // Connector SPI1 CS  
+    ConnectorName_SPI1_MOSI = Px_xx,  // Connector SPI1 MOSI  
+    ConnectorName_SPI1_MISO = Px_xx,  // Connector SPI1 MISO  
+    ConnectorName_SPI1_SCK  = Px_xx,  // Connector SPI1 SCK  
+    .  
+    .  
+    .  
+    
+    ConnectorName_SPIN_CS   = Px_xx,  // Connector SPIN CS  
+    ConnectorName_SPIN_MOSI = Px_xx,  // Connector SPIN MOSI  
+    ConnectorName_SPIN_MISO = Px_xx,  // Connector SPIN MISO  
+    ConnectorName_SPIN_SCK  = Px_xx,  // Connector SPIN SCK  
+
+### Board Components Pins
+
+**LED Definition**
+
+All LED definitions should be defined as follows:
+
+    // Px_xx relates to the processor pin connected to the LED
+    
+    LED1 = Px_xx,  // LED1  
+    LED2 = Px_xx,  // LED2  
+    LED3 = Px_xx,  // LED3  
+    LED4 = Px_xx,  // LED4  
+    .  
+    .  
+    LEDN = Px_xx   // LEDN
+
+**BUTTON Definition**
+
+All BUTTON definitions should be defined as follows:
+
+    // Px_xx relates to the processor pin connected to the Button  
+    BUTTON1 = Px_xx,  // BUTTON1  
+    BUTTON2 = Px_xx,  // BUTTON2  
+    BUTTON3 = Px_xx,  // BUTTON3  
+    BUTTON4 = Px_xx,  // BUTTON4   
+    .  
+    .  
+    BUTTONN = Px_xx   // BUTTONN  
+
+# Other information
+
+### Avoid Unnecessary Definitions
+
+If only 1 LED/BUTTON is on board, do not define additional LEDs/BUTTONs in BSP. This allows for unavailable LEDs/BUTTONs to be caught as a compile error.
```

Comment:
Is that what we've done up to now? I don't think so.

The downside of not defining there's no way to conditionally cope if you do that. You can't `#ifdef` an enum value. But if they're defined as NC, you can test for them:

     if (BUTTON2 != NC) {
          // code for button 2
     }

That is effectively a compile-time if. In C++17 you could get fancy and force that with `if constexpr (BUTTON2 != NC)`.

If you want a compile error, you can force one with

    static_assert(BUTTON1 != NC && BUTTON2 != NC, "This application needs two buttons");

Otherwise you will get a run-time error when you try to use an NC pin.

Reply:
This was one of the problems, static_assert is a good one to be provided here.

Reply:
So why do we keep theses pin name in the enum?
and not use define ?

Reply:
Legacy convention. You could change it, but currently `#ifdef BUTTON1` is false on all platforms, so tells you nothing. If every platform was universally updated, then that construct would be useful.

Another minor(ish) issue is that defines can be more intrusive. Defining something short like `A0` can cause more collision upset than having an `A0` identifier.

I've seen cases like

    template <typename A0>

which explode if you have

    #define A0 (PIN_H0)   // PIN_H0 is enum value

but you do get away with

    #define A0 PIN_H0

because `template <typename PIN_H0>` still works. If `PIN_H0` was `#define`d as something, and not an enum value, then you'd be in trouble.

Reply:
Other solution, is to move alias definition into json config file?

Reply:
I think it's clear that we need `NA`.
The actual question is what to do with that when the application uses that, then we can figure out implementation details. Options:
- Ignore (silent compile)
- Warning
- Compile error